### PR TITLE
Guided Remediation: Add manifest resolution

### PR DESCRIPTION
**Pull request by michaelkedar**
Starting to make guided remediation public #352 :tada:  

This PR has the code used to resolve the dependency graph of a manifest (i.e. a `package.json`) using the deps.dev resolvers and find vulnerabilities within it.

Doesn't include the code to actually parse/write `package.json` files - will probably add that in the next PR.

Much of this has been reviewed internally already, but I've made some significant changes/refactoring to `dependency_chain.go` and the `computeVulns()` function in `resolve.go`, so please take a more careful look at those.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,9 +1,12 @@
 module github.com/google/osv-scanner
 
-go 1.21
+go 1.21.1
+
+toolchain go1.21.5
```
Comment (michaelkedar) [inline go.mod:5]: Hm, `go mod tidy` changes these automatically - what should I be doing about this?

Comment (oliverchang) [reply go.mod:5]: @another-rex  ?

Comment (another-rex) [reply go.mod:5]: Update it to 1.21.5, I think when dependencies start adding in patch versions in their go.mod file, it forces it on us as well when we import it.

**Review by oliverchang — APPROVED** [review]

File: internal/resolution/client/depsdev_client.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+				System:  pb.System(vk.System),
+				Name:    vk.Name,
+				Version: vk.Version,
+			},
+		})
+		if err != nil {
+			continue
+		}
+
+		// Send off queries to cache the packages in the dependency tree
```
Comment (another-rex) [inline internal/resolution/client/depsdev_client.go:56]: Is this caching remote (on deps.dev servers) or local (guessing from datasource/depsdev_api.go)?  Trying to figure out how this caches packages.

**Review by cuixq — APPROVED** [review]
🎉

File: internal/resolution/datasource/depsdev_api.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+)
+
+// DepsDevAPIClient is a wrapper for InsightsClient that caches requests.
+type DepsDevAPIClient struct {
+	pb.InsightsClient
+
+	// cache fields
+	mu                sync.Mutex
+	cacheTimestamp    *time.Time
+	packageCache      map[packageKey]*pb.Package
```
Comment (another-rex) [inline internal/resolution/datasource/depsdev_api.go:22]: Is it possible to `sync.Map` instead of locking and unlocking here?

**Review by another-rex — COMMENTED** [review]
What's a good way to test this manually before we get automated tests?

File: internal/resolution/client/depsdev_client.go (hunk 2 above)
Comment (michaelkedar) [reply internal/resolution/client/depsdev_client.go:56]: This is local - The [deps.dev implementation of the client](https://github.com/google/deps.dev/blob/main/util/resolve/api.go) calls the relevant `datasource/depsdeb_api.go` methods, which will fill in its own maps.

**Comment by oliverchang** [comment]
> What's a good way to test this manually before we get automated tests?

Re tests, it would be easiest to run the internal version for now. Check with @michaelkedar on this.

File: internal/resolution/datasource/depsdev_api.go (hunk 3 above)
Comment (michaelkedar) [reply internal/resolution/datasource/depsdev_api.go:22]: I *could*, it's just annoying having to cast from `any` every time when using it.
Plus it's probably better to have a single lock for the `GobEncode` / `GobDecode` functions

Comment (another-rex) [reply internal/resolution/datasource/depsdev_api.go:22]: Fair enough, let's stick with lock then!

**Comment by michaelkedar** [comment]
Yeah, manually testing this as-is would be clumsy - you have to manually fill in the `Manifest` struct.

**Review by another-rex — APPROVED** [review]

**Pull request merged**